### ISSUE TITLE
Crash in nativeOptimizeNIR in fastparse/classparse test suite

### ISSUE THREAD

**Issue by lihaoyi** [scala-native/scala-native#1057]
```
lihaoyi Workspace$ git clone https://github.com/lihaoyi/fastparse
Cloning into 'fastparse'...
remote: Counting objects: 9664, done.
remote: Compressing objects: 100% (147/147), done.
remote: Total 9664 (delta 50), reused 246 (delta 41), pack-reused 9371
Receiving objects: 100% (9664/9664), 12.00 MiB | 10.50 MiB/s, done.
Resolving deltas: 100% (3565/3565), done.

lihaoyi Workspace$ cd fastparse/

lihaoyi fastparse$ git rev-parse head
c8ab9b6a72a5b6cae88e918de156e0adce85396d

lihaoyi fastparse$ sbt classparseNative/test
[info] Loading project definition from /Users/lihaoyi/Dropbox/Workspace/fastparse/project
[info] Updating {file:/Users/lihaoyi/Dropbox/Workspace/fastparse/project/}fastparse-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[warn] Found version conflict(s) in library dependencies; some are suspected to be binary incompatible:
[warn]
[warn] 	* org.scala-native:sbt-crossproject:0.2.2 is selected over {0.2.1, 0.1.0}
[warn] 	    +- org.scala-native:sbt-scalajs-crossproject:0.2.2 (scalaVersion=2.10, sbtVersion=0.13) (depends on 0.2.2)
[warn] 	    +- org.scala-native:sbt-scala-native:0.3.3 (scalaVersion=2.10, sbtVersion=0.13) (depends on 0.1.0)
[warn] 	    +- default:fastparse-build:0.1-SNAPSHOT (scalaVersion=2.10, sbtVersion=0.13) (depends on 0.2.1)
[warn]
[warn] Run 'evicted' to see detailed eviction warnings
[info] Compiling 1 Scala source to /Users/lihaoyi/Dropbox/Workspace/fastparse/project/target/scala-2.10/sbt-0.13/classes...
[info] Resolving key references (15385 settings) ...
[info] Set current project to fastparse (in build file:/Users/lihaoyi/Dropbox/Workspace/fastparse/)
[info] Compiling 7 Scala sources to /Users/lihaoyi/Dropbox/Workspace/fastparse/utils/native/target/scala-2.11/classes...
[warn] there was one deprecation warning; re-run with -deprecation for details
[warn] one warning found
[info] Compiling 15 Scala sources to /Users/lihaoyi/Dropbox/Workspace/fastparse/fastparse/native/target/scala-2.11/classes...
[warn] /Users/lihaoyi/Dropbox/Workspace/fastparse/fastparse/shared/src/main/scala/fastparse/parsers/Combinators.scala:347: abstract type R in type pattern fastparse.parsers.Combinators.Sequence[R,R,R,Elem,Repr] is unchecked since it is eliminated by erasure
[warn]           case p: Sequence[R, R, R, Elem, Repr] =>
[warn]                   ^
[warn] there were 6 feature warnings; re-run with -feature for details
[warn] two warnings found
[info] Compiling 3 Scala sources to /Users/lihaoyi/Dropbox/Workspace/fastparse/fastparseByte/native/target/scala-2.11/classes...
[warn] there were two feature warnings; re-run with -feature for details
[warn] one warning found
[info] Compiling 3 Scala sources to /Users/lihaoyi/Dropbox/Workspace/fastparse/classparse/native/target/scala-2.11/classes...
[warn] /Users/lihaoyi/Dropbox/Workspace/fastparse/classparse/shared/src/main/scala/classparse/ClassParse.scala:70: abstract type pattern T is unchecked since it is eliminated by erasure
[warn]             case info: T => Some(info)
[warn]                        ^
[warn] /Users/lihaoyi/Dropbox/Workspace/fastparse/classparse/shared/src/main/scala/classparse/ClassAttributes.scala:62: match may not be exhaustive.
[warn] It would fail on the following inputs: ClassInfo(_), FieldRefInfo(_, _), InterfaceMethodRefInfo(_, _), InvokeDynamicInfo(_, _), MethodHandleInfo(_, _), MethodRefInfo(_, _), MethodTypeInfo(_), NameAndTypeInfo(_, _), NopInfo(), Utf8Info(_)
[warn]       classInfo.getInfoByIndex[PoolInfo](idx).get match {
[warn]                                               ^
[warn] /Users/lihaoyi/Dropbox/Workspace/fastparse/classparse/shared/src/main/scala/classparse/ClassParse.scala:176: match may not be exhaustive.
[warn] It would fail on the following inputs: BasicElemInfo(_), ClassInfo(_), InvokeDynamicInfo(_, _), MethodHandleInfo(_, _), MethodTypeInfo(_), NameAndTypeInfo(_, _), NopInfo(), StringInfo(_), Utf8Info(_)
[warn]           classInfo.getInfoByIndex[PoolInfo](info.referenceIndex).get match {
[warn]                                                                   ^
[warn] three warnings found
[info] Compiling 2 Scala sources to /Users/lihaoyi/Dropbox/Workspace/fastparse/classparse/native/target/scala-2.11/test-classes...
[info] Compiling 1 Scala source to /Users/lihaoyi/Dropbox/Workspace/fastparse/classparse/native/target/scala-2.11/test-classes...
[info] Linking (3334 ms)
[info] Discovered 2662 classes and 18346 methods
java.util.NoSuchElementException: key not found: Block(Local(28),Vector(),UnrolledBuffer(Let(Local(86),Classalloc(Top(scala.MatchError))), Let(Local(87),Box(Class(Top(java.lang.Integer)),Local(Local(25),Int))), Let(Local(88),Call(Function(Vector(Class(Top(scala.MatchError)), Class(Top(java.lang.Object))),Unit),Global(Member(Top(scala.MatchError),init_java.lang.Object),Ptr),Vector(Local(Local(86),Class(Top(scala.MatchError))), Local(Local(87),Class(Top(java.lang.Integer)))),None)), Throw(Local(Local(86),Class(Top(scala.MatchError))),None)),false)
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:58)
	at scala.collection.MapLike$class.apply(MapLike.scala:141)
	at scala.collection.AbstractMap.apply(Map.scala:58)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination.isDominatedBy$1(GlobalBoxingElimination.scala:40)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination.scala$scalanative$optimizer$pass$GlobalBoxingElimination$$canReuse$1(GlobalBoxingElimination.scala:45)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination$$anonfun$onInsts$1$$anonfun$apply$1.isDefinedAt(GlobalBoxingElimination.scala:57)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination$$anonfun$onInsts$1$$anonfun$apply$1.isDefinedAt(GlobalBoxingElimination.scala:54)
	at scala.collection.TraversableOnce$$anonfun$collectFirst$1.apply(TraversableOnce.scala:132)
	at scala.collection.TraversableOnce$$anonfun$collectFirst$1.apply(TraversableOnce.scala:131)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.collectFirst(TraversableOnce.scala:131)
	at scala.collection.AbstractTraversable.collectFirst(Traversable.scala:105)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination$$anonfun$onInsts$1.apply(GlobalBoxingElimination.scala:54)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination$$anonfun$onInsts$1.apply(GlobalBoxingElimination.scala:51)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination.onInsts(GlobalBoxingElimination.scala:51)
	at scala.scalanative.optimizer.Pass$class.onDefn(Pass.scala:33)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination.onDefn(GlobalBoxingElimination.scala:16)
	at scala.scalanative.optimizer.Pass$$anonfun$onDefns$1.apply(Pass.scala:22)
	at scala.scalanative.optimizer.Pass$$anonfun$onDefns$1.apply(Pass.scala:22)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.mutable.UnrolledBuffer$Unrolled.foreach(UnrolledBuffer.scala:224)
	at scala.collection.mutable.UnrolledBuffer.foreach(UnrolledBuffer.scala:130)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at scala.scalanative.optimizer.Pass$class.onDefns(Pass.scala:22)
	at scala.scalanative.optimizer.pass.GlobalBoxingElimination.onDefns(GlobalBoxingElimination.scala:16)
	at scala.scalanative.optimizer.Optimizer$.scala$scalanative$optimizer$Optimizer$$loop$1(Optimizer.scala:64)
	at scala.scalanative.optimizer.Optimizer$$anonfun$apply$1.apply(Optimizer.scala:74)
	at scala.scalanative.optimizer.Optimizer$$anonfun$apply$1.apply(Optimizer.scala:70)
	at scala.collection.parallel.AugmentedIterableIterator$class.map2combiner(RemainsIterator.scala:120)
	at scala.collection.parallel.immutable.ParHashMap$ParHashMapIterator.map2combiner(ParHashMap.scala:79)
	at scala.collection.parallel.ParIterableLike$Map.leaf(ParIterableLike.scala:1057)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
	at scala.collection.parallel.ParIterableLike$Map.tryLeaf(ParIterableLike.scala:1054)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:183)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:162)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinTask.doJoin(ForkJoinTask.java:341)
	at scala.concurrent.forkjoin.ForkJoinTask.join(ForkJoinTask.java:673)
	at scala.collection.parallel.ForkJoinTasks$WrappedTask$class.sync(Tasks.scala:444)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.sync(Tasks.scala:514)
	at scala.collection.parallel.ForkJoinTasks$class.executeAndWaitResult(Tasks.scala:492)
	at scala.collection.parallel.ForkJoinTaskSupport.executeAndWaitResult(TaskSupport.scala:64)
	at scala.collection.parallel.ParIterableLike$ResultMapping.leaf(ParIterableLike.scala:961)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
	at scala.collection.parallel.ParIterableLike$ResultMapping.tryLeaf(ParIterableLike.scala:956)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:165)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[error] (classparseNative/nativetest:nativeOptimizeNIR) java.util.NoSuchElementException: key not found: Block(Local(28),Vector(),UnrolledBuffer(Let(Local(86),Classalloc(Top(scala.MatchError))), Let(Local(87),Box(Class(Top(java.lang.Integer)),Local(Local(25),Int))), Let(Local(88),Call(Function(Vector(Class(Top(scala.MatchError)), Class(Top(java.lang.Object))),Unit),Global(Member(Top(scala.MatchError),init_java.lang.Object),Ptr),Vector(Local(Local(86),Class(Top(scala.MatchError))), Local(Local(87),Class(Top(java.lang.Integer)))),None)), Throw(Local(Local(86),Class(Top(scala.MatchError))),None)),false)
[error] Total time: 47 s, completed 14 Oct, 2017 10:08:29 AM
lihaoyi fastparse$
```

**Comment by densh** [comment]
Hi @lihaoyi, thanks for the report and the details instructions. The issue is clearly reproducible.

**Comment by densh** [comment]
Fixed in #1058
